src/linker/Linker.Steps/DiscoverCustomOperatorsHandler.cs

```diff
@@ -0,0 +1,215 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public class DiscoverOperatorsHandler : IMarkHandler
+	{
+		LinkContext _context;
+		bool seenLinqExpressions;
+		readonly HashSet<TypeDefinition> _trackedTypesWithOperators;
+		Dictionary<TypeDefinition, HashSet<MethodDefinition>> _pendingOperatorsForType;
```

Comment:
I would use `Dictionary<TypeDefinition, List<MethodDefinition>>` as far as I can tell we don't do `Contains` checks on the value collection, we only enumerate over it. So hash set is not appropriate (and a tiny bit more expensive). And we should never try to add a duplicate to that list since we only process each type once.